src/manager/checkpoint.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use std::collections::hash_map::RandomState;
+use std::collections::{HashMap, HashSet};
+use std::time::Duration;
+
+use anyhow::Result;
+use cid::Cid;
+use fil_actors_runtime::cbor;
+use futures_util::future::join_all;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::MethodNum;
+use ipc_gateway::Checkpoint;
+use ipc_sdk::subnet_id::SubnetID;
+use primitives::TCid;
+use tokio::time::sleep;
+
+use crate::config::Subnet;
+use crate::lotus::client::LotusJsonRPCClient;
+use crate::lotus::message::mpool::MpoolPushMessage;
+use crate::lotus::LotusClient;
+
+const CHAIN_HEAD_REQUEST_PERIOD: Duration = Duration::from_secs(10);
+
+/// Starts the checkpoint manager.
+#[allow(dead_code)]
+pub async fn start(subnets: HashMap<String, Subnet>) {
+    let managed_subnets = subnets_to_manage(subnets);
+
+    let mut futures = Vec::new();
+    for (child, parent) in managed_subnets {
+        futures.push(manage_subnet(child, parent));
+    }
+    join_all(futures).await;
+}
+
+/// This function takes a `HashMap<String, Subnet>` and returns a `Vec` of tuples of the form
+/// `(child_subnet, parent_subnet)`, where `child_subnet` is a subnet that we need to actively
+/// manage checkpoint for. This means that for each `child_subnet` there exists at least one account
+/// for which we need to submit checkpoints on behalf of to `parent_subnet`, which must also be
+/// present in the map.
+fn subnets_to_manage(subnets: HashMap<String, Subnet>) -> Vec<(Subnet, Subnet)> {
+    // First, we remap subnets by SubnetID.
+    let subnets_by_id: HashMap<SubnetID, Subnet> = subnets
+        .values()
+        .map(|s| (s.id.clone(), s.clone()))
+        .collect();
+
+    // Then, we filter for subnets that have at least one account and for which the parent subnet
+    // is also in the map, and map into a Vec of (child_subnet, parent_subnet) tuples.
+    subnets_by_id
+        .values()
+        .filter(|s| !s.accounts.is_empty())
+        .filter(|s| s.id.parent().is_some() && subnets_by_id.contains_key(&s.id.parent().unwrap()))
+        .map(|s| (s.clone(), subnets_by_id[&s.id.parent().unwrap()].clone()))
+        .collect()
+}
+
+/// Monitors a subnet `child` for checkpoint blocks. It emits an event for every new checkpoint block.
+async fn manage_subnet(child: Subnet, parent: Subnet) -> Result<()> {
+    let child_client = LotusJsonRPCClient::from_subnet(&child);
+    let parent_client = LotusJsonRPCClient::from_subnet(&parent);
+
+    // Read the parent's chain head and obtain the tip set CID.
+    let parent_head = parent_client.chain_head().await?;
+    let cid_map = parent_head.cids.last().unwrap().clone();
```

Comment:
nit: can we get `first()` here? we shouldn't ever see more than one block, but in certain parts of the lotus code we use `block[0]`, that way if something goes wrong we are sure that we are being consistent with the lotus logic.

Reply:
I can do that, of course. But isn't the last block the one that represents the state we want to check for?

Reply:
But this list of CIDs is the list of blocks for the heaviest height, right? So for one block per tipset `first() == last()` or am I wrong?

Reply:
My previous comment was a way of asking whether we should change this to `first()` or change lotus to use `last()` (or equivalent). If the last block in the tipset is the one that represents the state we're considering, then it should be `last()`. In any case, change this here for now.